dev/AppLifecycle/AppInstance.cpp

```diff
@@ -0,0 +1,415 @@
+        if (data == nullptr)
+        {
+            data = args;
+        }
+        
+        return { kind, data };
+    }
+
+    AppInstance::AppInstance(uint32_t processId)
+    {
+        m_processId = processId;
+        m_isCurrent = (GetCurrentProcessId() == processId);
+
+        m_moduleName = ComputeAppId();
+        m_processName = wil::str_printf<std::wstring>(L"%s_%d", m_moduleName.c_str(), processId);
+
+        m_instances.Init(m_moduleName + L"_Module");
+
+        // Wire up the Activated event.
+        std::wstring eventName = m_processName + c_activatedEventNameSuffix;
+        m_innerActivated.create(wil::EventOptions::ManualReset, eventName.c_str());
+
+        // This mutex should always be created first by the process it's associated with.
+        std::wstring mutexName = m_processName + L"_Mutex";
+        m_dataMutex.create(mutexName.c_str(), CREATE_MUTEX_INITIAL_OWNER, MUTEX_ALL_ACCESS);
+        auto alreadyExists = (GetLastError() == ERROR_ALREADY_EXISTS);
+
+        // This means the AppInstance object was constructed for a process that
+        // does not exist.  This is typical for when we found an orphan in m_instances.
+        auto orphan = (!alreadyExists && !m_isCurrent);
+        assert(!orphan);
+
+        wil::mutex_release_scope_exit releaseOnExit;
+        auto weak_this = get_weak();
+        if (m_isCurrent)
+        {
+            releaseOnExit = m_dataMutex.ReleaseMutex_scope_exit();
+
+            auto onInnerActivated = [weak_this]
+            {
+                // If this weak ref doesn't resolve it means the process is heading toward a terminal state.
+                // If that happens just ignore any redirections as they can't be serviced.
+                auto strong_this{ weak_this.get() };
+                if (strong_this)
+                {
+                    strong_this->ProcessRedirectionRequests();
+                }
+            };
+
+            m_activationWatcher.create(m_innerActivated.get(), onInnerActivated);
+        }
+        else
+        {
+            releaseOnExit = m_dataMutex.acquire();
+        }
+
+        m_key.Open(m_processName + L"_Key");
+
+        if (m_isCurrent)
+        {
+            m_instances.Insert(processId);
+        }
+        else
+        {
+            m_instanceHandle.reset(OpenProcess(SYNCHRONIZE, FALSE, processId));
+
+            // Create a monitor thread to handle cleaning up this instance if the backing process terminates.
+            auto onInstanceTerminated = [weak_this]
+            {
+                auto strong_this{ weak_this.get() };
+                if (strong_this)
+                {
+                    strong_this->OnInstanceTerminated();
+                }
+            };
+
+            m_terminationWatcher.create(m_instanceHandle.get(), onInstanceTerminated);
+        }
+
+        m_redirectionArgs.Init(m_processName + L"_RedirectionQueue");
+    }
+
+    void AppInstance::OnInstanceTerminated()
```

Comment:
>AppInstance::OnInstanceTerminated() [](http://example.com/codeflow?start=9&length=35)

I see _reactive_ behavior, removing PIDs when the API is called. Is there a _proactive_ monitoring for process death to cleanup on process termination? That would reduce the window for PID recycling to grief you